gpui: Add SVG rendering to `img` element

MatthiasGrandl opened the pull request:
Release Notes:

- Updated dependencies `tiny_skia`, `usvg`, `resvg`
- Added SVG rendering to `img` element


Up for discussion:
~~I implemented the rasterization of the SVG in the img.rs, because I wanted to make sure it gets rasterized at the correct size everytime. This is probably too expensive and should be optimized still. Not sure how to approach this yet.~~


Edit:
To add some reasoning as to why I think SVG rendering support belongs into the img element:

- The `svg` element seems to be specifically for embedded and monochrome SVGs.
- Having SVG support in the `img` element allows GPUI to be closer to the HTML img tag.

User Story:
While working on the [Loungy](https://github.com/MatthiasGrandl/loungy) Linux port, we discovered that app icons are either PNGs or SVGs. Having the ability to display both in the same element would save us lots of code and additional dependencies.

MatthiasGrandl commented:
> I implemented the rasterization of the SVG in the img.rs, because I wanted to make sure it gets rasterized at the correct size everytime. This is probably too expensive and should be optimized still. Not sure how to approach this yet.

I pushed a change that caches each rendered size of an SVG, so that SVGs aren't rasterized every frame.

mikayla-maki commented on crates/gpui/src/elements/img.rs:
```diff
@@ -210,12 +210,77 @@ impl Element for Img {
+                                            )
+                                            .unwrap();
+                                            let ratio = width as f32 / tree.size().width();
+                                            resvg::render(
```
It's not good to do a whole SVG render while constructing elements. We should move this entire block into the future generated by `image_cache`

mikayla-maki replied:
If we need to pass the bounds in to `ImageCache::get`, that's fine.

MatthiasGrandl replied:
> If we need to pass the bounds in to ImageCache::get, and just ignore it in most other cases, that's fine.

turns out that's not enough since I need a lot more than just the bounds. I also need the scale factor and the object-fit to calculate the final size of the rendered image. that's obviously a bit ugly

mikayla-maki commented on crates/gpui/src/image_cache.rs:
```diff
@@ -63,7 +72,16 @@ impl From<Arc<PathBuf>> for UriOrPath {
+#[derive(Clone)]
+pub enum FetchImageType {
+    Raster(Arc<ImageData>),
+    Vector {
```
We should remove this type and have the SVG rasterization generate an ImageData.

MatthiasGrandl replied:
agreed, makes sense.

MatthiasGrandl replied:
done, let me know what you think

mikayla-maki commented:
I think this whole approach is a bit incorrect. We already have SVG rendering and caching going on with the `svg()` element. Could we adjust that type to accept in-memory data (not just a path) and then delegate the implementation in `img` to `svg`?

MatthiasGrandl commented:
> I think this whole approach is a bit incorrect. We already have SVG rendering and caching going on with the `svg()` element. Could we adjust that element to accept in-memory data (not just a path) and then delegate the implementation in `img` to `svg`?

@mikayla-maki 
That seems a bit daunting. As I said in the description of this PR, the `svg` element seems to be purpose built for monochrome icons. If you insist, I can refactor the `svg` element to be more generic, but since the SVG rendering logic is rather tiny, I think it doesn't hurt to separate.

mikayla-maki commented:
> The svg element seems to be purpose built for monochrome icons.

Fair point, if this needs to have non-monochrome rendering, then splitting out this implementation separately is fine. Either way though, we need to get rendering out of the hot loop and into a future :)

MatthiasGrandl committed:
gpui: Add SVG rendering to img element

MatthiasGrandl committed:
gpui: move SVG rendering further down

MatthiasGrandl commented:
@mikayla-maki I basically started from scratch. I think most of your valid criticisms should be handled now. This new approach also has the benefit that users can create both SVG and Raster ImageData with the `try_from_bytes` function, which is nice as then you don't need to pull in the `image` crate if you don't need it elsewhere.

niklaswimmer commented on crates/gpui/Cargo.toml:
```diff
@@ -66,8 +66,8 @@ sum_tree.workspace = true
 time.workspace = true
-tiny-skia = "0.5"
-usvg = { version = "0.14", features = [] }
+tiny-skia = "0.11"
+usvg = { version = "0.40", features = [] }
```
Just wanted to note that resvg reexports both tiny-skia and usvg, so there is no need for us to depend on those directly and risking to get the versions out of sync by accident.

MatthiasGrandl replied:
Yes I am aware. I can gladly change it. Just thought maybe there was a reason for the dependencies to be listed independently.

MatthiasGrandl replied:
Done

MatthiasGrandl committed:
gpui: remove unnecessary dependencies

MatthiasGrandl commented:
@mikayla-maki @niklaswimmer anything else you want changed? I am pretty happy with the state of this PR

niklaswimmer commented on crates/gpui/src/assets.rs:
```diff
@@ -35,40 +43,135 @@ impl AssetSource for () {
+            let data = resvg::usvg::Tree::from_data(
+                bytes,
+                &resvg::usvg::Options::default(),
+                &resvg::usvg::fontdb::Database::default(),
```
This will create an empty font database. I am not sure what exactly usvg uses it for tho - it may be fine. Alternatively we can create a database that contains the system fonts https://docs.rs/fontdb/latest/fontdb/struct.Database.html#method.load_system_fonts and pass that in here.

MatthiasGrandl replied:
Ok yeah that definitely makes sense. Otherwise SVGs with text embedded just won’t display the text. Do you think we need to cache the fontdb? No idea how expensive it is to initialize it for each SVGs.

niklaswimmer replied:
Depends on how many fonts the user has on their system. I think caching would be better, since it needs to read quite a few files every time it loads the fonts.

MatthiasGrandl replied:
@niklaswimmer do you think a OnceLock would be acceptable for caching the fontdb? I am still quite new to Rust so just want to make sure we are not endlessly iterating on some minor detail.

MatthiasGrandl committed:
gpui: cache populated fontdb for svg rendering

MatthiasGrandl committed:
add image_cache documentation

MatthiasGrandl commented:
I rebased the pull request, because there have been changes to the `img` element and added documentation because apparently clippy rules have changed.

mikayla-maki commented:
Hi @MatthiasGrandl, 

I talked it over with some other folks on the Zed team and I don't think we want to go with the approach in this PR. SVG rendering is just too different from image rendering. That said, I think we can take the same code from this PR and incorporate it differently:


1. Add an API to the `svg` element, roughly like so:
 ```rs
pub fn source(mut self, source: impl Into<VectorSource>) -> Self {
        self.source = Some(source.into());
        self
    }
```
2. Add an `svgs` field to the ImageCache, that can provide a mapping like `HashMap<(VectorSource, Size), FetchVectorDataTask>`. I don't expect SVG icons to have their size changing very often, so I don't think we need to cache both yet. Expose this through an API similar to the existing `ImageCache::get()`.
3. Adjust the `paint_svg()` call to take and use a `VectorData`, that can be constructed from a `VectorSource` type
4. With these two changes, your use case can be achieved like so:

```rs
// psuedo code
let icon = if icon_file.is_image() {
    img().source(icon_file).into_any()
} else {
   svg().source(icon_file).into_any()
}
```

The details around the representation of `ImageSource` and `VectorSource` seem a little tricky, particularly because of the inclusion of the `Surface` type, which should not be a part of the `img` element at all IMO. But I think most of the code you've written here can be repurposed for this style of API. If you want to take a stab at separating out a `surface` element, and then unifying the `ImageSource` and `VectorSource` APIs into some kind of `DataSource`, feel free. But I don't think that will be necessary to ship this feature.

MatthiasGrandl commented:
@mikayla-maki I disagree with this as a GPUI user. Having to differentiate between SVG and raster images yourself is super unergonomic.

Your pseudo code means you have to manually implement a cache, because without having the files in memory, you can’t differentiate between filetypes. It also mean you will need to pull both `image` and `resvg` in all GPUI apps that need this.

Also SVGs aren’t exclusive to Icons, they can be used like any image so I disagree with the assessment that you don’t need a size specific cache,

I don’t think I will be working on this. I am already using the current code in Loungy.

mikayla-maki commented:
Good point on the image caching, I had only considered the caching internal to GPUI and not external to it. I'll give it more thought :)

MatthiasGrandl commented:
@mikayla-maki I'll gladly have a discussion with the team around this PR. I really do think it's important to have compatibility for SVGs in the `img` element. It is kinda expected in the world of HTML. And this PR works really well.

I also really need this to get merged if I don't want to permanently maintain a GPUI fork. It's impossible to replicate this functionality in the Loungy codebase (because I don't have the render bounds to calculate the correct SVG size).

mikayla-maki commented:
The two issues I had with this PR the first time where that:
- GPUI's `paint_image()` call should not have to handle the ImageCache's implementation details, like the `now_or_never` call
- The ImageData implementation depended on the implementation of `paint_image`

 I've gone through a few alternative designs, and I think I found one which should work for the constraints here. Essentially, we should generalize how GPUI loads and processes assets so that we can move all the complexity for handling both raster and vector images into the `img` element implementation. This would allow us to write code like so:

```rs
// in the `img` element's `impl Element` block
fn before_layout(/**/) {
    cx.with_cache::<RasterOrVector>(&self.source, |asset, cx| {
       // Use the asset.size()
   })
}
//...
fn paint(bounds) {
   cx.with_cache::<RasterOrVector>(&self.source, |asset, cx| {
       match asset {
           RasterOrVector::Raster { image_data, ..} {
                 cx.paint_image(image_data, /**/);
           }
           RasterOrVector::Vector { tree } {
                 let key = VectorCacheKey { tree, size: bounds.size };
                 cx.with_cache::<Vector>(&key, |vector, cx| {
                     cx.paint_image(vector.image_data, /**/);
                 })   
           }
      }

   })
}
```

Some notes:
1. `RasterOrVector` would be similar to the `ImageData` type that's currently in this PR, but moved into the `img` code. 
2. `paint_image()` can work largely as it already does, and just pass the old `ImageData` straight through.
3. `with_cache()` is the key interface here, and it needs two supporting traits:

```rs
trait Asset {
    type Key: Hash;
    type Output;
    async fn load(source: &Self::Source, cx: &mut AppContext) -> Self::Output;
}
```

RasterOrVector would implement this interface like so:

```
impl Asset for RasterOrVector {
    type Key = UriOrPath;
    type Output = Self;

    async fn load(source: &UriOrPath, cx: &mut AppContext) -> Self {
       // This would be largely the same as `imageCache::get()` + `ImageData::try_from_bytes()`
    }
}
```

And `Vector` would have an implementation like so: 


```
struct VectorKey {
  tree: usvg::Tree,
  bounds: Size<DevicePixels>,
}

// Implement Hash for this type, ignore the `tree` field and just use the bounds for the hash

impl Asset for Vector {
    type Key = VectorKey;
    type Output = ImageData;   

    async fn load(source: &VectorKey, cx: &mut AppContext) -> Self::Output {
       // This would be largely the same as the code currently in`ImageData::rendered()`
    }
}
```

With this supporting type infrastructure, the `with_cache()` call can take on the general details of cache management. It should probably be backed by a `HashMap<(TypeId, u64), Shared<Task<Box<dyn Any>>>`, and it should do the `now_or_never` + `cx.on_next_frame(|cx| cx.refresh())` details that `ImageSource::data` is currently doing. 

The exact details of how these traits look will probably change when it's actually being implemented. But I think doing this will move the complexity of supporting both types of images out of the main GPUI world and into the `img` element :)

Thoughts @MatthiasGrandl ?

mikayla-maki commented:
Notes from discord:

We should change this from `with_cache` to `with_asset`. This simplifies and narrows the API, allowing caching to be handled as an internal concern to the asset's loading function. We can still provide a general `AssetCache` on GPUI, that works like the ImageCache does today. But the `with_asset` API won't use it. It'll just handle the future management + scheduling another frame to be rendered once the asset is loaded. This allows easy pass-through implementations, where the GPUI user provides it's own in-memory data.

mikayla-maki commented:
Given the significant divergence from the original implementation of this PR, I'm going to close this PR for now. Let's make a new one with the `with_asset` approach :)
